File: service1/service1.py
import os
import subprocess
import requests
import json
import time
import signal
import sys
from flask import Flask, jsonify
from flask_cors import CORS

app = Flask(__name__)
CORS(app)

def handle_shutdown(signum, frame):
    sys.exit(0)

signal.signal(signal.SIGTERM, handle_shutdown)

# Function to get system information
def get_system_info():
    ip_address = os.popen('hostname -I').read().strip()
    processes = os.popen('ps -ax').read()
    disk_space = os.popen('df -hP /').read()
    uptime = os.popen('uptime -p').read().strip()
    
    return {
        'ip_address': ip_address,
        'processes': processes,
        'disk_space': disk_space,
        'uptime': uptime
    }

sleeping = False
def intermittent_sleep():
    global sleeping
    sleep_time = 2
    if not sleeping:
        sleeping = True
        print(f"Sleeping for {sleep_time} seconds...")
        time.sleep(sleep_time)
        print(f"{sleep_time} seconds have passed")
        sleeping = False

@app.after_request
def sleep_after_response(response):

    @response.call_on_close
    def sleep_after_close():
        intermittent_sleep()

    return response

@app.route('/info', methods=['GET'])
def get_info():
    # Get information from Service1
    service1_info = get_system_info()

    # Request information from Service2
    try:
        service2_response = requests.get('http://service2:5001/')
        service2_info = service2_response.json()
    except requests.RequestException as e:
        service2_info = {"error": "Could not reach Service2", "details": str(e)}

    # Combine information from both services
    result = {
        'Service1': service1_info,
        'Service2': service2_info
    }

    print("Returning response from service1")

    return jsonify(result)

@app.route('/shutdown', methods=['POST'])
def shutdown():
    try:
        subprocess.run(["docker", "compose", "down"], check=True)
        return jsonify({"message": "System is shutting down..."})
    except subprocess.CalledProcessError as e:
        return jsonify({"error": f"Failed to shut down: {e}"}), 500

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8199)

File: service1/service1.js
const express = require('express');
const axios = require('axios');
const os = require('os');
const { exec } = require('child_process');

const app = express();
const PORT = 8199;

// Cooldown period control
let isCoolingDown = false;

// Helper function to execute shell commands and return output
const execCommand = (command) => {
    return new Promise((resolve, reject) => {
        exec(command, (error, stdout, stderr) => {
            if (error) {
                reject(stderr);
            } else {
                resolve(stdout);
            }
        });
    });
};

// Function to get system information
async function getSystemInfo() {
    const ipAddress = (await execCommand('hostname -I')).trim();
    const processes = await execCommand('ps -ax');
    const diskSpace = await execCommand('df -hP /');
    const uptime = (await execCommand('uptime -p')).trim();

    return {
        ip_address: ipAddress,
        processes: processes,
        disk_space: diskSpace,
        uptime: uptime
    };
}

app.get('/info/', async (req, res) => {
    if (isCoolingDown) {
        res.status(503).json({ error: 'Service is temporarily unavailable due to cooldown period' });
        return;
    }

    // Set cooldown to true to block further requests
    isCoolingDown = true;

    try {
        const service1Info = await getSystemInfo();

        // Request information from Service2
        let service2Info = {};
        try {
            const response = await axios.get('http://service2:5001/');
            service2Info = response.data;
        } catch (error) {
            service2Info = { error: 'Failed to connect to Service2' };
        }

        // Combine and send response
        res.json({
            Service1: service1Info,
            Service2: service2Info
        });

        // Cooldown period (2 seconds)
        setTimeout(() => {
            isCoolingDown = false;
        }, 2000);
    } catch (error) {
        res.status(500).json({ error: 'Internal server error' });
        isCoolingDown = false; // Reset cooldown in case of error
    }
});

app.listen(PORT, () => {
    console.log(`Service1 is running on port ${PORT}`);
});
